### PR TITLE
AUTH-440: OIDC: refresh sessions with a refresh token if present

### PR DESCRIPTION
**Pull request by stlaz**
This PR significantly improves user experience in environments with short-lived id-tokens (which is typical for OIDC, in fact).

The PR introduces a *combined session handler*:
- stores id-tokens *server-side* as these can easily grow over maximum allowed cookie size
- stores refresh tokens *client-side*, encrypted by gorilla.CookieStore
- stores session tokens *client-side*, these session tokens point to the server-side session with the id-token

If a server-side stored id-token is expiring/expired, an attempt to refresh it will be made if a refresh token is present in the client-side session. This helps us avoid having to redirect the user to their OIDC every time their ID token expires, which can be as often as once every minute!

~Currently only a single session is supported among instances. This should be improved in a follow-up.~
Allows multiple parallel sessions, one per instance.

/assign @jhadvig 
/cc @TheRealJon
/cc @spadgett

### CODE REVIEW

**Comment by jhadvig** [comment]
QE Approver:
/assign @yapei
Docs Approver:
/assign @opayne1
PX Approver:
/assign @RickJWagner

**Comment by yapei** [comment]
@yanpzhan will work on this

File: pkg/auth/auth_openshift.go [hunk 1]
```diff
@@ -124,14 +120,11 @@ func (o *openShiftAuth) getOIDCDiscoveryInternal(ctx context.Context) (*oidcDisc
 	return metadata, nil
 }
 
-func (o *openShiftAuth) login(w http.ResponseWriter, token *oauth2.Token) (*loginState, error) {
+func (o *openShiftAuth) login(w http.ResponseWriter, _ *http.Request, token *oauth2.Token) (*sessions.LoginState, error) {
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/auth/auth_openshift.go:123]: it's the interface

File: pkg/auth/auth_oidc.go [hunk 2]
```diff
@@ -50,48 +52,54 @@ func newOIDCAuth(ctx context.Context, sessionStore *SessionStore, c *oidcConfig)
...
 	}
-	var c json.RawMessage
-	if err := idToken.Claims(&c); err != nil {
-		return nil, fmt.Errorf("parsing claims: %v", err)
+	if err := o.sessions.AddSession(w, r, ls); err != nil {
+		return nil, err
 	}
-	ls, err := newLoginState(rawIDToken, []byte(c))
+
+	o.sessions.PruneSessions()
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/auth/auth_oidc.go:72]: it's possible pruning is broken. I'll have to verify.
I'll later open a PR that does pruning in background and contains fixes if necessary.

File: pkg/auth/auth_oidc.go [hunk 3]
```diff
@@ -102,61 +110,49 @@ func (o *oidcAuth) verify(ctx context.Context, rawIDToken string) (*oidc.IDToken
...
-		return nil, fmt.Errorf("No session found on server")
+		return nil, fmt.Errorf("failed to retrieve login state: %v", err)
 	}
-	if ls.exp.Sub(ls.now()) < 0 {
-		o.sessions.deleteSession(sessionToken)
-		return nil, fmt.Errorf("Session is expired.")
+
+	if ls == nil || ls.IsExpired() {
+		if refreshToken := o.sessions.GetCookieRefreshToken(r); refreshToken != "" {
+			return o.refreshSession(r.Context(), w, r, o.oauth2Config(), refreshToken)
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/auth/auth_oidc.go:131]: we could refine this for certain network errors in a follow up

**Comment by yanpzhan** [comment]
I'm testing on cluster launched against the pr. Regression test about normal idp passed. I tried to login console using external OIDC by deploying keycloak, but met error in console url "https://console-openshift-console.apps.yanpzhtest.qe.devcluster.openshift.com/auth/error?error=Invalid+scopes%3A+user%3Afull&error_type=auth". And there is error info in keycloak pod logs:
`2024-01-26 08:30:29,328 INFO  [org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory] (main) Node name: node_393506, Site name: null
2024-01-26 08:30:29,441 INFO  [org.keycloak.broker.provider.AbstractIdentityProviderMapper] (main) Registering class org.keycloak.broker.provider.mappersync.ConfigSyncEventListener
2024-01-26 08:30:29,497 INFO  [org.keycloak.services] (main) KC-SERVICES0050: Initializing master realm
2024-01-26 08:30:31,843 INFO  [io.quarkus] (main) Keycloak 23.0.4 on JVM (powered by Quarkus 3.2.9.Final) started in 14.585s. Listening on: http://0.0.0.0:8080
2024-01-26 08:30:31,844 INFO  [io.quarkus] (main) Profile dev activated. 
2024-01-26 08:30:31,844 INFO  [io.quarkus] (main) Installed features: [agroal, cdi, hibernate-orm, jdbc-h2, jdbc-mariadb, jdbc-mssql, jdbc-mysql, jdbc-oracle, jdbc-postgresql, keycloak, logging-gelf, micrometer, narayana-jta, reactive-routes, resteasy-reactive, resteasy-reactive-jackson, smallrye-context-propagation, smallrye-health, vertx]
2024-01-26 08:30:32,178 INFO  [org.keycloak.services] (main) KC-SERVICES0009: Added user 'admin' to realm 'master'
2024-01-26 08:30:32,188 WARN  [org.keycloak.quarkus.runtime.KeycloakMain] (main) Running the server in development mode. DO NOT use this configuration in production.
2024-01-26 08:53:51,980 WARN  [org.keycloak.services] (executor-thread-7) KC-SERVICES0091: Request is missing scope 'openid' so it's not treated as OIDC, but just pure OAuth2 request.
2024-01-26 08:53:51,980 ERROR [org.keycloak.services] (executor-thread-7) KC-SERVICES0093: Invalid parameter value for: scope
2024-01-26 08:53:51,982 WARN  [org.keycloak.events] (executor-thread-7) type=LOGIN_ERROR, realmId=974fd686-6195-4856-af4f-2fce0329991f, clientId=console, userId=null, ipAddress=120.244.160.198, error=invalid_request, response_type=code, redirect_uri=https://console-openshift-console.apps.yanpzhtest.qe.devcluster.openshift.com/auth/callback, response_mode=query
2024-01-26 08:54:00,696 ERROR [org.keycloak.services] (executor-thread-7) KC-SERVICES0093: Invalid parameter value for: scope
2024-01-26 08:54:00,697 WARN  [org.keycloak.events] (executor-thread-7) type=LOGIN_ERROR, realmId=974fd686-6195-4856-af4f-2fce0329991f, clientId=console, userId=null, ipAddress=209.132.188.14, error=invalid_request, response_type=code, redirect_uri=https://console-openshift-console.apps.yanpzhtest.qe.devcluster.openshift.com/auth/callback, response_mode=query
2024-01-26 08:54:06,841 ERROR [org.keycloak.services] (executor-thread-7) KC-SERVICES0093: Invalid parameter value for: scope
2024-01-26 08:54:06,841 WARN  [org.keycloak.events] (executor-thread-7) type=LOGIN_ERROR, realmId=974fd686-6195-4856-af4f-2fce0329991f, clientId=console, userId=null, ipAddress=209.132.188.14, error=invalid_request, response_type=code, redirect_uri=https://console-openshift-console.apps.yanpzhtest.qe.devcluster.openshift.com/auth/callback, response_mode=query
`

**Comment by stlaz** [comment]
The above errors were observed because the operator never got to deploy the proper config revision, this is getting fixed in https://github.com/openshift/console-operator/pull/849

**Comment by yanpzhan** [comment]
Tested on cluster launched against the pr, after deployed keycloak and did some configuration, accessed console url, on the keycloak login page, input user/password, console page opened successfully. Observed the tokens in developer console, tokens could refresh and console page keeps open steadily.
/label qe-approved

**Pull request merged**